산책 도장 리스트 개선및 프로그래스바 게이지 초기화 시점 변경

IamJuco opened the pull request:
- 산책 도장 리스트 개선
- 프로그래스바 게이지 초기화 시점을 최근 1주일이 아닌 월요일 기준으로 변경
## 참고사항
commit issue number를 잘 못 표기한게 있습니다. 유의바랍니다!!


resolved: #242

IamJuco committed:
add: Stamp 목록을 불러올 모델 생성

issue: #208

IamJuco committed:
Merge branch 'develop' into feature/walk-stamp-list

IamJuco committed:
Merge branch 'develop' into feature/walk-stamp-list

IamJuco committed:
add: 도장 버튼용 shape drawable 추가

issue: #208

IamJuco committed:
design: 도장 전체리스트 아이템 수정

issue: #208

IamJuco committed:
design: 도장 전체리스트 아이템 수정및 strings 수정

이제 안쓰는 strings삭제및 수정, 디자인 도장 전체 리스트로 수정

issue: #208

IamJuco committed:
feat: 산책 도장 목록 전체보기 완료

usecase를 getStampInfoListUseCase로 변경

issue: #208

IamJuco committed:
design: 클릭 디자인 수정

issue: #208

IamJuco committed:
feat: 도장 설명을 위한 data와 domain layer 수정

issue: #208

IamJuco committed:
feat: stampInfo, description 수정

issue: #208

IamJuco committed:
add: 모든 산책도장 리스트용 fragment생성

issue: #208

IamJuco committed:
rename: 용도에맞게 파일명 수정

issue: #208

IamJuco committed:
feat: 바인딩처리

issue: #208

IamJuco committed:
rename: xml파일 용도에맞게 이름 수정

issue: #208

IamJuco committed:
refactor: Activity를 Fragment로 전환

AllStampFragment에서 화면을 보여줘야하는 이벤트를 한번 더 해야하기에 리소스를 잡아먹는 Activity대신 Fragment로 전환

issue: #208

IamJuco committed:
add: 산책도장 디테일화면을 위한 fragment추가

issue: #208

IamJuco committed:
feat: 클릭이벤트 초기작업및 Stamp관련 모델을 StampModel로 변경

클릭이벤트 초기 코드만 작성 하고, Stamp모델을 StampModel과 StampListModel로 두개사용하던것을 하나로 통합

issue: #208

IamJuco committed:
remove: 안쓰는 모델 삭제

Stamp모델을 하나로 통합하여 쓰지않는 StampListModel 삭제

issue: #208

IamJuco committed:
feat: 데이터 넘기는 클릭 이벤트 완성

issue: #208

IamJuco committed:
design: 산책 도장 디테일화면 초기 디자인

issue: #208

IamJuco committed:
dosc: 알아보기쉽게 모델 설명, 주석 처리

issue: #208

IamJuco committed:
style: 라인정렬

issue: #208

IamJuco committed:
style: text, Strings로 변경

issue: #208

IamJuco committed:
add: detail용 아이콘 추가

issue: #208

IamJuco committed:
add: stamp detail용 아이템 추가

issue: #208

IamJuco committed:
design: 디테일 디자인 완성

issue: #208

IamJuco committed:
add: 산책도장용 어댑터추가

issue: #208

IamJuco committed:
add: Detail viewmodel 추가

issue: #208

IamJuco committed:
feat: Adapter및 리싸이클러뷰 아이디 설정

issue: #208

IamJuco committed:
feat: model에 description 추가

issue: #208

IamJuco committed:
feat: 사용할 유스케이스 변경

issue: #208

IamJuco committed:
feat: StampDetail 기능 완성

issue: #208

IamJuco committed:
fix: 산책도장을 제대로 못가져오던 버그 해결

산책도장을 각 산책도장마다 똑같이 갯수를 가져오던 버그 수정

issue: #208

IamJuco committed:
add: recyclerview background color 변경을위한 drawable 추가

issue: #208

IamJuco committed:
design: 디자인 수정

텍스트뷰 크기, 배경색 화이트로 변경등 수정

issue: #208

IamJuco committed:
feat: back 버튼 추가

issue: #208

IamJuco committed:
feat: 날짜 형식 변경및 함수 이름 수정

날짜 형식수정및 여러곳에서 쓰이는 함수이기에 용도에맞게 함수 이름 수정

issue: #208

IamJuco committed:
Merge branch 'develop' into feature/walk-stamp-list

IamJuco committed:
design: 스탬프 도장 흭득여부 텍스트 추가

issue: #208

IamJuco committed:
feat: 산책 도장 데이터 유무에따른 예외처리

issue: #208

IamJuco committed:
style: 라인정렬

issue: #208

IamJuco committed:
Merge branch 'develop' into feature/walk-stamp-list

IamJuco committed:
feat: 산책도장 흭득 시점 수정

산책 도장을 얻는기준이 1주일마다 초기화가 되어, 기존에 최근 1주일 마다 산책 도장이 업데이트되는 시점을 월요일부터 일요일까지로 변경

issue: #242

neoneoneo123 commented on app/src/main/java/com/nbcfinalteam2/ddaraogae/presentation/ui/stamp/AllStampAdapter.kt:
```diff
@@ -17,23 +17,25 @@ class AllStampAdapter : ListAdapter<StampModel, AllStampAdapter.ViewHolder>(DIFF
 
     companion object {
         private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<StampModel>() {
             override fun areItemsTheSame(oldItem: StampModel, newItem: StampModel): Boolean {
-                return oldItem.id == newItem.id
+                return oldItem.num == newItem.num
             }
```
저는 DiffUtil에서 areItemsTheSame 함수 안에서는 아이템들의 **고유값**을 비교하는 것으로 이해하고 있습니다.
(https://jermainedilao.medium.com/demystifying-diffutil-itemcallback-class-8c0201cc69b1)

사용하시는 stampModel의 num 값의 역할이 고유값의 역할을 하는 걸까요?
id에서 num으로 변경하신 이유가 무엇인지 궁금하네요!

IamJuco replied:
네! 처음 AllStampAdapter를 구성할때는 그저 사용자가 얻은 산책 도장의 총 갯수를 가져오는 화면을 구성 하였었는데요.
 당시에는 GetStampListByPeriodUseCase를 이용하였는데 이 usecase는 StampEntity를 return해주고 있었습니다.

    @Parcelize
    data class StampEntity(
        val id: String?,
        val stampNum: Int?,
        val getDateTime: Date?,
        val name: String?,
    ) : Parcelable
그래서 id 값을 비교해주고 있었는데,

산책 리스트 화면을 개선하면서 GetStampListByPeriodUseCase에서 GetStampInfoListUseCase를 사용하게 되었는데
해당 유스케이스의 return값은 StampInfoEntity로 변수 num이 고유 id 값을 대신 하고 있었습니다.
        
    data class StampInfoEntity(
        val num: Int,
        val title: String,
        val description: String
    )

종합하면 사용하는 usecase를 바꾸게 되었는데 바꾼 usecase에서 return값 중에 고유값의 명칭이 id에서 num으로 변경되어 바꾸게 되었습니다.

neoneoneo123 replied:
아 명칭이 바뀐걸로 이해하면 되는거군요. 알겠습니다!

neoneoneo123 commented on app/src/main/java/com/nbcfinalteam2/ddaraogae/presentation/ui/stamp/AllStampFragment.kt:
```diff
@@ -0,0 +1,96 @@
+import javax.inject.Inject
+
+@AndroidEntryPoint
+class AllStampFragment : Fragment() {
```
<img width="840" alt="image" src="https://github.com/nbc-final-team2/ddaraogae/assets/128209823/5ff0512b-6c6c-48b1-97af-2a4ccea750e9">

화면을 회전하면 리스트의 아이템들이 좌측으로 몰려보이는 현상이 있습니다!

IamJuco replied:
해당 내용 확인하고 수정 하였습니다.
item에서 부모 컨테이너를 wrap_content로 했더니 문제가 발생하였네용
match_parent로 변경하면서 해결하였습니다!
e3a6a09b332d9fcb572396b06479014f0a2e70db

추가로 Click하면 넘어가는 화면에서도 가로모드 방어를 위해 NestedScrollView를 추가하였습니다!
4c4c38ae6ca08c73c1025f21c431a1f1ad20e2f6

neoneoneo123 reviewed (changes requested):
성현님, 고생 많으셨습니다!

스탬프 상세 내용을 확인할 수 있어서 좋은 것 같아요.

2가지 정도의 코멘트가 있어서 남겨드립니다!

IamJuco committed:
fix: 가로모드시 아이템들이 왼쪽으로 몰리던 버그 수정

width를 wrap_content로 하여 크기가 정해져있어 왼쪽으로 몰리던현상을 match_Parent로 해결

issue: #242

IamJuco committed:
fix: 가로모드방어

recyclerview가 있어 nestedScrollview로 혹시 모를 가로모드에 대한 방어

issue: #242

IamJuco committed:
Merge branch 'develop' into feature/walk-stamp-list

IamJuco committed:
fix: invisible 시점 변경

초기값을 invisible로 처리하여 조금 더 자연스럽게 뷰가 보이도록 수정

issue: #242

IamJuco committed:
refactor: back 버튼로직, lifecycleScope 형식 변경및 flow처리

back버튼을 deprecated된것을 finish로 변경, viewlifecycleScope는 뷰가 사라질때 데이터가 유실되어, lifecycleScope로 로직변경및 flow처리

issue: #242

winteryy commented on app/src/main/java/com/nbcfinalteam2/ddaraogae/presentation/ui/stamp/StampDetailViewModel.kt:
```diff
@@ -0,0 +1,66 @@
+    private val getStampInfoListUseCase: GetStampInfoListUseCase
+) : ViewModel() {
+
+    private val _stampDetail = MutableLiveData<StampModel>()
+    val stampDetail: LiveData<StampModel> = _stampDetail
+
+    private val _stampList = MutableLiveData<List<StampModel>>()
+    val stampList: LiveData<List<StampModel>> = _stampList
```
StateFlow로 리팩터링 가능하신가요?

IamJuco replied:
네! 한번 시도해 보았습니다.

93e4b7108171fe1648d4ce48d9e342f553e4ec4c

winteryy commented on app/src/main/java/com/nbcfinalteam2/ddaraogae/presentation/ui/stamp/StampDetailFragment.kt:
```diff
@@ -0,0 +1,71 @@
+        stampDetailViewModel.loadStampDetail(stampId)
+    }
+
+    private fun setupListener() {
+        binding.btnBack.setOnClickListener {
+            requireActivity().onBackPressed()
+        }
+    }
```
어제 언급했던 부분인데 아직 안 고쳐진 것 같습니당

IamJuco replied:
확인했고 수정했습니다~

04e96983b562e8a668dff45d61ebefaba5e6667a

theBettor reviewed (approved):
지연님 피드백으로 개선하신 부분까지 확인했습니다 고생하셨어요!

winteryy commented on app/src/main/res/layout/fragment_stamp_detail.xml:
```diff
@@ -0,0 +1,125 @@
+    tools:context=".presentation.ui.stamp.StampDetailFragment">
+
+    <androidx.core.widget.NestedScrollView
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
```
ConstraintLayout 내부 요소인데 match_parent로 되어 있네용

IamJuco replied:
확인했고 수정했습니다~

e1840cb7ba142afee5765f9cb8e4895ef91eecc4

winteryy reviewed (changes requested):
나머지는 어제 zep에서 말씀드렸던 부분이라 고쳐진 것 같습니다
몇 부분만 확인 부탁드립니다!

IamJuco committed:
refactor: back버튼 로직 finish로 변경

issue: #242

IamJuco committed:
fix: Constraintlayout 내부요소 match -> 0dp로변경

issue: #242

IamJuco committed:
refactor: livedate -> StateFlow로 변경

issue: #242

winteryy reviewed (approved):
고생 많으셨습니다👍👍
